### PR TITLE
Has a binary messenger

### PR DESCRIPTION
**Pull request by gaaclarke**
I made FlutterViewController and FlutterEngine both have binary messengers instead of publicly being binary messengers.  This breaks certain retain cycles that we could create.  It also minimizes the damage of the current problem we have where channels always leak on iOS.  It has a fail-safe mechanism where talking to dead channels just gets ignored.

This won't break plugins, but it will potentially break some add2app scenarios.  The client may have been saying:
```
[FlutterMethodChannel methodChannelWithName:@"foo" binaryMessenger:flutterViewController]
```

They will now have to say:
```
[FlutterMethodChannel methodChannelWithName:@"foo" 
                            binaryMessenger:flutterViewController.binaryMessenger]
```

### CODE REVIEW

**Comment by gaaclarke** [comment]
https://flutter.dev/docs/development/platform-integration/platform-channels#step-4a-add-an-ios-platform-specific-implementation-using-objective-c

This example code will have to change as a result of this PR.

**Comment by gaaclarke** [comment]
Relevant issue: https://github.com/flutter/flutter/issues/26007

**Comment by gaaclarke** [comment]
Email has been sent to flutter-announce: https://groups.google.com/forum/#!topic/flutter-announce/ewwf2Rhm9Uo

File: shell/platform/darwin/ios/framework/Headers/FlutterViewController.h [hunk 1]
```diff
@@ -165,6 +164,12 @@ FLUTTER_EXPORT
  */
 @property(weak, nonatomic, readonly) FlutterEngine* engine;
 
+/**
+ * The `FlutterBinaryMessenger` associated with this FlutterViewController (used for communicating
+ * with channels).
+ */
+@property(nonatomic, readonly) NSObject<FlutterBinaryMessenger>* binaryMessenger;
```
Comment (xster) [inline shell/platform/darwin/ios/framework/Headers/FlutterViewController.h]: If we're going to break the API, we might as well not leak the implementation detail knowledge of the FlutterEngine (which happens to have a binaryMessenger) into the FlutterViewController.

File: shell/platform/darwin/macos/framework/Source/FLETextInputPlugin.mm [hunk 2]
```diff
@@ -72,7 +72,7 @@ - (instancetype)initWithViewController:(FLEViewController*)viewController {
   if (self != nil) {
     _flutterViewController = viewController;
     _channel = [FlutterMethodChannel methodChannelWithName:kTextInputChannel
-                                           binaryMessenger:viewController
+                                           binaryMessenger:viewController.binaryMessenger
```
Comment (xster) [inline shell/platform/darwin/macos/framework/Source/FLETextInputPlugin.mm]: can we just do viewController.engine.binaryMessenger here? (assuming the lifecycle sequence is right)

File: shell/platform/darwin/ios/framework/Headers/FlutterViewController.h (hunk 1 above)
Comment (gaaclarke) [reply shell/platform/darwin/ios/framework/Headers/FlutterViewController.h]: From just looking at FlutterViewController you have no idea you are talking with the engine.  I think we could remove it but it's a slightly more complicated migration.  I tried to be as easy as possible.  If we wanted to go further I'd recommend a separate PR in case someone raises an objection to that change.

File: shell/platform/darwin/macos/framework/Source/FLETextInputPlugin.mm (hunk 2 above)
Comment (gaaclarke) [reply shell/platform/darwin/macos/framework/Source/FLETextInputPlugin.mm]: I think we could.  I'm not too confident in the change considering I don't think we have alot of test coverage on this code.  I'd prefer doing the simplest, slightest correct to get what we want.

**Comment by gaaclarke** [comment]
@xster Sounds good wrt Android.  I haven't looked at the Android side much.  We probably need a bigger exploration to see how we can unify.  I poked around a bit and they are pretty diverged right now so I don't know if its necessary to have to make an android change in response to this change yet.

**Comment by xster** [comment]
yes, there's unfortunately a bit of divergence currently. Since it's a breaking change, we should ideally plan ahead a bit and not break a second time this quarter. I pinged you a list of other API differences. Let's chat tomorrow to work out something progressive but that we won't break in the future.

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm [hunk 3]
```diff
@@ -498,6 +503,11 @@ - (void)performAction:(FlutterTextInputAction)action withClient:(int)client {
   return _shell->Screenshot(type, base64Encode);
 }
 
+#pragma mark - FlutterBinaryMessengerContainer
```
Comment (xster) [inline shell/platform/darwin/ios/framework/Source/FlutterEngine.mm]: This may have been refactored but FlutterBinaryMessengerContainer doesn't seem to refer to anything concrete.

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm [hunk 4]
```diff
@@ -14,16 +14,20 @@
...
+#import "flutter/shell/platform/darwin/ios/framework/Source/FlutterPlatformPlugin.h"
+#import "flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputDelegate.h"
+#import "flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.h"
+#import "flutter/shell/platform/darwin/ios/framework/Source/FlutterView.h"
+#import "flutter/shell/platform/darwin/ios/framework/Source/platform_message_response_darwin.h"
+#import "flutter/shell/platform/darwin/ios/platform_view_ios.h"
 
 NSNotificationName const FlutterSemanticsUpdateNotification = @"FlutterSemanticsUpdate";
 
+@interface FlutterViewController () <FlutterBinaryMessenger>
```
Comment (xster) [inline shell/platform/darwin/ios/framework/Source/FlutterViewController.mm]: Why do we need this? Aren't we already breaking the public API? Why not just delete everything under #pragma mark - FlutterBinaryMessenger?

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm [hunk 5]
```diff
@@ -50,6 +55,7 @@ - (instancetype)initWithEngine:(FlutterEngine*)engine
   NSAssert(engine != nil, @"Engine is required");
   self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
   if (self) {
+    _binaryMessenger = [[FlutterBinaryMessengerRelay alloc] initWithParent:self];
```
Comment (xster) [inline shell/platform/darwin/ios/framework/Source/FlutterViewController.mm]: should this be initWithParent:engine.binaryMessenger?

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm [hunk 6]
```diff
@@ -989,23 +998,28 @@ - (void)onPreferredStatusBarStyleUpdated:(NSNotification*)notification {
   return [_engine.get() platformViewsController];
 }
 
+#pragma mark - FlutterBinaryMessengerContainer
+- (NSObject<FlutterBinaryMessenger>*)binaryMessenger {
```
Comment (xster) [inline shell/platform/darwin/ios/framework/Source/FlutterViewController.mm]: I'd deprecate this right now as we mentioned and encourage the vc.engine.binaryMessenger path like we mentioned.

File: shell/platform/darwin/macos/framework/Source/FLEViewController.mm [hunk 7]
```diff
@@ -70,7 +70,7 @@ void Reset() {
 /**
  * Private interface declaration for FLEViewController.
  */
-@interface FLEViewController ()
+@interface FLEViewController () <FlutterBinaryMessenger>
```
Comment (xster) [inline shell/platform/darwin/macos/framework/Source/FLEViewController.mm]: These ones have engines too right? Can we add the relaying mechanism here too since we don't want the desktop view to power the channels either (could be in a different PR).

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm (hunk 4 above)
Comment (gaaclarke) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm]: (discussed offline)  This is the private API so that the Relay knows he can just forward to the view controller..

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm (hunk 5 above)
Comment (gaaclarke) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm]: No, that would create a relay on top of a relay.  This is how we want it.  Optionally we could remove binaryMessenger from the ViewController header, and/or make it just return the engines binaryMessenger.  I recommend doing neither since it is the easiest migration and the safest change.

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm (hunk 6 above)
Comment (gaaclarke) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm]: I can't actually deprecate this =(.  The accessibility bridge requires the indirection to behave correctly:  https://github.com/flutter/engine/blob/185087a65f5104710399d107a5228636b558727c/shell/platform/darwin/ios/framework/Source/accessibility_bridge.mm#L575

File: shell/platform/darwin/macos/framework/Source/FLEViewController.mm (hunk 7 above)
Comment (gaaclarke) [reply shell/platform/darwin/macos/framework/Source/FLEViewController.mm]: (discussed offline) Yep, we should do the same thing for desktop.  I refrained from doing it now because the way the build system is setup it isn't easy to share the code currently.  This should present no change to the current functionality of the code while putting in place the API we want.

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm (hunk 6 above)
Comment (gaaclarke) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm]: Done.  I just remembered why it had this dependency and removed it, it shouldn't be needed anymore (and especially after we remove the leak completely).

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm (hunk 3 above)
Comment (gaaclarke) [reply shell/platform/darwin/ios/framework/Source/FlutterEngine.mm]: Done, removed here and in the ViewController.
